### ISSUE TITLE
SuperSource Transitions!

### ISSUE THREAD

**Issue by davidjoshuaford** [bitfocus/companion-module-bmd-atem#144]
While the action: _**"atem: Change SuperSource box properties"**_ is great at CUTTING between SuperSource Box parameters, 

AND **_"Offset SuperSource box properties"_** command is useful in MANUALLY ADJUSTING SuperSource Box, 

    ---> there's currently no way to TRANSITION between SuperSource parameters. 

I imagine this would involve some algorithm that considers where each box is currently placed, and then dynamically calculates keyframes over time, to transition the boxes to the new parameters. 

Is this a possible enhancement? 

Use cases: when the SuperSource is already on program, Transitions are necessary to avoid jump cuts when modifying the layouts.

Many thanks!
David

<img width="525" alt="Screen Shot 2021-06-03 at 10 32 05 AM" src="https://user-images.githubusercontent.com/76959770/120662900-848d8c80-c457-11eb-8594-6b8eb3165be3.png">

**Comment by Julusian** [comment]
In theory this would be pretty simple to implement, as it is already being done in this module for audio.
However, doing it for video will require a much higher tickrate to appear smooth (for audio we are by default at 10 per second). Also I suspect that companion might cause glitching when it decides to do some other things, so it might end up not being useful

**Comment by davidjoshuaford** [comment]
Would Companion have the resources to be able to run a 1/3 or 1/2-second transition? I don't think it would need to be longer than that. And even low-resolution key-frame rate (ie 5 to 10 keyframes within 500ms) would be OK. I've been playing with manual transitions (moving objects off-screen using repetitions of the "offset" properties action), and there's enough compute power to get a box across the screen.

**Comment by Julusian** [comment]
Its impossible to say. Its entirely dependant on the machine, and which other modules are running (some will be worse partners than others). Also, on what else you do with the atem.

I have suspected for a while that some of the updates from commands that the atem module does are strangely slow. Ive just checked, and on my machine I am getting a 38ms+ time taken. This is currently when input properties, or media player assets change so rare but not unheard of.
At 60p, each frame is 16.6ms, so best case scenario is will freeze for 2-3 frames. 

And this is just from the atem module. I have no idea if other operations inside other modules will be just as slow or slower. Because all the modules share a single thread, any one of them doing a long/slow block of code will block others from executing.
And it will likely work 90% of the time, but the remaining time will stutter unexpectedly

**Comment by davidjoshuaford** [comment]
By "slow ATEM module commands", do you mean simple things like changing preview and program inputs take longer? If so, I'd say I've noticed a slight decrease in responsiveness in the builds that have been released the last month or two. Nothing that would be show stopping, but there is a slight lag compared to switching via the ATEM software. 

This is my rudimentary testing of transitioning boxes off screen. The first few frames run ok, but the tail end gets a bit choppy. And companion is definitely losing some commands along the way, because it's hit or miss whether the box returns to the starting position. 

https://user-images.githubusercontent.com/76959770/121087089-fb0cf000-c7b1-11eb-93a4-a616c985617f.mov

**Comment by Julusian** [comment]
No, I mean commands that require a lot of processing to handle the state updates. Whenever input properties change, we have to regenerate all the actions, feedbacks, presets and variables because they all use the input names in their options. Its somehow surprisingly slow work (I need to do some digging to figure out why)

**Comment by davidjoshuaford** [comment]
Ok, sounds tricky! Look forward to seeing how you go with it. Thank you!

**Comment by davidjoshuaford** [comment]
Hi @Julusian, Been thinking through SS again – since it's easy to implement, would it be possible to give it a go and we experiment with how fluid it is on different systems? Cheers, David

**Comment by davidjoshuaford** [comment]
Hi @Julusian, with modules being run in child mode in version 3 of Companion ahead of us, would that present more compute power & reliability to calculate Supersource keyframe transitions?

**Comment by Julusian** [comment]
Yeah, I think this will be a lot more reliable now, so is something to explore

**Comment by davidjoshuaford** [comment]
Wonderful! That would be a huge win :) 

The other killer feature to add alongside this would be a "learn" action (instead of a mouse-click). Just like we can do with the mouse in the GUI for the ATEM's supersource inputs.....

<img width="637" alt="Screenshot 2022-12-15 at 11 01 12 AM" src="https://user-images.githubusercontent.com/76959770/207909201-f47cf060-bf89-4d83-ad6f-abfe6bb2d3aa.png">

...if this could be a "Learn Supersource Parameters" Action that were assigned to a long-press, then we have a function that is similar to what I use for Save & Recall (long & short press respectively) on PTZ cameras.... or similar to how car radio stations let you recall a station with a short press, and save the current station with a long press...

**Comment by Julusian** [comment]
> The other killer feature to add alongside this would be a "learn" action (instead of a mouse-click).

The big problem is identifying the action to run learn on. In a dropdown list to select the action, should it show `atem: SuperSource: Set box source` or something else? 
Maybe just a number field for an index, but also consider the 'step' and 'group' the action is in will need to be specified.

If it is just done for other actions on the same button, then this doesn't get too complicated to do overall, as we will have all the actions loaded, so can produce a good list. If you want it for an action on another button, then we can't really suggest anything.

**Comment by davidjoshuaford** [comment]
I imagine we'd want to learn approx as many as 10 supersource actions on a single button – ie learn the four box parameters, background/foreground art settings, and the four input sources. Hence would there be a general companion internal action that would "learn parameters on this button"? For instance, if that "learn" action was added to a long press, it would read and apply all actions from the short press section (effectively like it was clicking all the "learn" buttons of each action, but not actually executing the action).

**Comment by davidjoshuaford** [comment]
> In theory this would be pretty simple to implement, as it is already being done in this module for audio. However, doing it for video will require a much higher tickrate to appear smooth (for audio we are by default at 10 per second). Also I suspect that companion might cause glitching when it decides to do some other things, so it might end up not being useful

Hi Julian, circling back on supersource transitions - would it be easy enough to start by implementing transitions like has been done for audio fades? I imagine 15 tickrate over a half second would be enough for a minimum default.
